packages/shared/src/contexts/PushNotificationContext.tsx

```diff
@@ -0,0 +1,150 @@
+  isInitialized: boolean;
+  isSubscribed: boolean;
+  isLoading: boolean;
+  shouldOpenPopup: boolean;
+  onSourceChange: (source: string) => void;
+  trackPermissionGranted: (source: NotificationPromptSource) => void;
+}
+
+export const PushNotificationsContext =
+  createContext<PushNotificationsContextData>({
+    isPushSupported: false,
+    isInitialized: true,
+    isSubscribed: false,
+    isLoading: false,
+    shouldOpenPopup: true,
+    trackPermissionGranted: null,
+    onSourceChange: null,
+    OneSignal: null,
+  });
+
+interface PushNotificationContextProviderProps {
+  children: ReactElement;
+  app: BootApp;
+}
+
+export function PushNotificationContextProvider({
+  children,
+  app,
+}: PushNotificationContextProviderProps): ReactElement {
+  const notificationSourceRef = useRef<string>();
+  const onSourceChange = useCallback((source) => {
+    notificationSourceRef.current = source;
+  }, []);
+  const subscriptionCallbackRef = useRef<SubscriptionCallback>();
+  const [isSubscribed, setIsSubscribed] = useState(false);
+  const { user } = useAuthContext();
+  const { trackEvent } = useAnalyticsContext();
+  const isEnabled = !!user && !isTesting;
+  const key = generateQueryKey(RequestKey.OneSignal, user);
+  const client = useQueryClient();
+  const {
+    data: OneSignalCache,
+    isFetched,
+    isLoading,
+  } = useQuery<typeof OneSignal>(
+    key,
+    async () => {
+      const osr = client.getQueryData<typeof OneSignal>(key);
+
+      if (osr) {
+        return osr;
+      }
+
+      await OneSignal.init({
+        appId: process.env.NEXT_PUBLIC_ONESIGNAL_APP_ID,
+        serviceWorkerParam: { scope: '/push/onesignal/' },
+        serviceWorkerPath: '/push/onesignal/OneSignalSDKWorker.js',
+      });
+
+      await OneSignal.login(user.id);
+
+      setIsSubscribed(OneSignal.User.PushSubscription.optedIn);
+
+      OneSignal.User.PushSubscription.addEventListener(
+        'change',
+        ({ current }) => {
+          setIsSubscribed(() => current.optedIn);
+          subscriptionCallbackRef.current?.(current.optedIn);
+        },
+      );
+
+      return OneSignal;
+    },
+    { enabled: isEnabled, ...disabledRefetch },
+  );
+
+  const isPushSupported =
+    !!globalThis.window?.Notification &&
+    OneSignal.Notifications.isPushSupported();
+
+  const trackPermissionGranted = useCallback(
+    (source) => subscriptionCallbackRef.current?.(true, source, true),
+    [],
+  );
+
+  useEffect(() => {
+    subscriptionCallbackRef.current = (
+      isSubscribedNew,
+      source,
+      existing_permission,
+    ) => {
+      if (isSubscribedNew) {
+        trackEvent({
+          event_name: AnalyticsEvent.ClickEnableNotification,
+          extra: JSON.stringify({
+            origin: source || notificationSourceRef.current,
+            permission: 'granted',
+            ...(existing_permission && { existing_permission }),
+          }),
+        });
+      }
+    };
+  }, [trackEvent]);
```

Comment:
Is this only to keep `subscriptionCallbackRef.current` in sync with `trackEvent` reference? If so we can just assign inline during render without useEffect.

Reply:
Right. We don't wait for anything, so we can just add this.